### ISSUE TITLE
'dependency has a new identifier' error with Webpack 3 and optimize.ModuleConcatenationPlugin.

### ISSUE THREAD

**Issue by jgerigmeyer** [mzgoddard/hard-source-webpack-plugin#142]
After upgrading to Webpack 3 and using the new [webpack.optimize.ModuleConcatenationPlugin()](https://medium.com/webpack/webpack-3-official-release-15fd2dd8f07b#c11e), the first build works fine but subsequent builds error with `Unhandled rejection Error: dependency has a new identifier`.

Relevant webpack config:

```js
    new webpack.optimize.ModuleConcatenationPlugin(),
    new HardSourceWebpackPlugin({
      cacheDirectory: path.join(__dirname, 'jscache/[confighash]'),
      recordsPath: path.join(__dirname, 'jscache/[confighash]/records.json'),
      configHash: webpackConfig => nodeObjectHash().hash(webpackConfig),
      environmentHash: {
        root: process.cwd(),
        directories: ['node_modules'],
        files: ['package.json']
      }
    }),
```

Output on repeat builds (absolute paths replaced with `[...]`):

```bash
Unhandled rejection Error: dependency has a new identifier
    at [...]/node_modules/hard-source-webpack-plugin/index.js:1094:26
    at factory ([...]/node_modules/webpack/lib/NormalModuleFactory.js:253:5)
    at applyPluginsAsyncWaterfall ([...]/node_modules/webpack/lib/NormalModuleFactory.js:99:14)
    at [...]/node_modules/tapable/lib/Tapable.js:204:11
    at NormalModuleFactory.params.normalModuleFactory.plugin ([...]/node_modules/webpack/lib/CompatibilityPlugin.js:52:5)
    at NormalModuleFactory.applyPluginsAsyncWaterfall ([...]/node_modules/tapable/lib/Tapable.js:208:13)
    at resolver ([...]/node_modules/webpack/lib/NormalModuleFactory.js:74:11)
    at [...]/node_modules/hard-source-webpack-plugin/index.js:1411:11
    at fromCache ([...]/node_modules/hard-source-webpack-plugin/index.js:1374:18)
    at [...]/node_modules/hard-source-webpack-plugin/index.js:1381:18
    at [...]/node_modules/hard-source-webpack-plugin/index.js:1390:12
    at [...]/node_modules/webpack/lib/NormalModuleFactory.js:64:5
    at applyPluginsAsyncWaterfall ([...]/node_modules/webpack/lib/NormalModuleFactory.js:246:4)
    at [...]/node_modules/tapable/lib/Tapable.js:204:11
    at NormalModuleFactory.<anonymous> ([...]/node_modules/hard-source-webpack-plugin/index.js:1485:39)
    at [...]/node_modules/tapable/lib/Tapable.js:206:14
```

Looking at the code where the error is being thrown (https://github.com/mzgoddard/hard-source-webpack-plugin/blob/master/index.js#L1091), the values being compared are:

`cacheDependency._resolvedModuleIdentifier` (absolute paths replaced with `[...]`):

```bash
[...]/node_modules/babel-loader/lib/index.js??ref--0-0![...]/static/js/app/accounts/auth.js [...]/node_modules/babel-loader/lib/index.js??ref--0-0![...]/static/js/app/accounts/login.js [...]/node_modules/babel-loader/lib/index.js??ref--0-0![...]/static/js/app/organizations.js [...]/node_modules/babel-loader/lib/index.js??ref--0-0![...]/static/js/app/projects.js [...]/node_modules/babel-loader/lib/index.js??ref--0-0![...]/static/js/app/accounts/register.js [...]/node_modules/babel-loader/lib/index.js??ref--0-0![...]/static/js/app/scenarios.js [...]/node_modules/babel-loader/lib/index.js??ref--0-0![...]/static/js/app/tplSetup.js [...]/node_modules/babel-loader/lib/index.js??ref--0-0![...]/static/js/app/app.js
```

`depModule.identifier()` (absolute paths replaced with `[...]`):

```bash
[...]/node_modules/babel-loader/lib/index.js??ref--0-0![...]/static/js/app/app.js
```

On the first build, or any build without `webpack.optimize.ModuleConcatenationPlugin()` enabled, the entire block (https://github.com/mzgoddard/hard-source-webpack-plugin/blob/master/index.js#L1053-L1098) is not executed.

**Comment by mzgoddard** [comment]
I see what's going on. That error is doing its job. hard-source can't do its job if a dependency resolves to a different module so it has to invalidate the module it was parsed from.

e.g.
Module A depends on `'./b'`. `'./b'` resolves to Module B.
In the next build `'./b'` resolves to Module BB. hard-source has to invalidate Module A so it can rebuild with an updated resolution.

To support the `ConcatenatedModule`, hard-source needs to find the original module the dependency resolved to that is now part of `ConcatenatedModule` so it can record the proper `_resolvedModuleIdentifier` for when it performs this check.

**Comment by mzgoddard** [comment]
I've dug deep enough into this to have figured out a solution. It'll be in two stages.

The first stage will handle the message currently appearing as an unhandled rejection and allow modules depending on ones in the ConcatenatedModule to be cached. Modules that get concatenated will not be cached in this. In this hard-source will continue to help with modules with commonjs or amd syntax, or es6 module syntax that are not concatenated. This will likely primarily benefit libraries with large numbers of files that fall into those categories.

The first stage should be available soon. It should only be blocked by needing to check that the other tests pass for webpack 3. (Last I checked some hard-source es6 module tests failed.)

The second stage will support caching the modules in the ConcatenatedModule. To support this, the details of how these modules are cached are different enough than previously cached modules to increase my interest in refactoring some of the code to provide **internal** plugin hooks that can handle the details appropriately. These **internal** hooks will at a later time be used by other plugins. (I only mention this as I'll be very likely to change the internal hooks until they stabilize. Use them at your own disgression. The internal hooks will be distinguished from public hooks by being prefixed with two dashes `--` like how private object variables are often prefixed with two underscores `__`.)

**Comment by mzgoddard** [comment]
#146 has been merged into the `0.5` prerelease line with support for caching modules inside the ConcatenatedModule. `0.5` is ready for use, I'm looking to do a few more changes to finalize that version.

**Comment by mzgoddard** [comment]
Published `0.5` last week with the mentioned ConcatenationPlugin support.
